donkeycar/parts/robocars_hat_ctrl.py

```diff
@@ -0,0 +1,345 @@
+                user_steering = self.fixSteering            
+
+        elif self.ch4Feature == self.AUX_FEATURE_OUTPUT_STEERING_TRIM :
+            if (abs(self.lastAux2 - self.inAux2)>0.5) :
+                if self.inAux2 > 0.5:
+                    self.fixOutputSteeringTrim = min(self.fixOutputSteeringTrim+self.cfg.ROBOCARSHAT_OUTPUT_STEERING_TRIM_INC,2000)
+                    mylogger.info("CtrlIn Fixed output steering set to {}".format(self.fixOutputSteeringTrim))
+                if self.inAux2 < -0.5:
+                    self.fixOutputSteeringTrim = max(self.fixOutputSteeringTrim-self.cfg.ROBOCARSHAT_OUTPUT_STEERING_TRIM_INC,1000)
+                    mylogger.info("CtrlIn Fixed output steering set to {}".format(self.fixOutputSteeringTrim))
+                self.sensor.setSteeringTrim (self.fixOutputSteeringTrim)            
+
+        elif self.ch4Feature == self.AUX_FEATURE_OUTPUT_STEERING_EXP :
+            if (abs(self.lastAux2 - self.inAux2)>0.5) :
+                if self.inAux2 > 0.5:
+                    self.fixOutputSteering = min(self.fixOutputSteering+self.cfg.ROBOCARSHAT_OUTPUT_STEERING_TRIM_INC,2000)
+                    mylogger.info("CtrlIn Fixed output steering set to {}".format(self.fixOutputSteering))
+                if self.inAux2 < -0.5:
+                    self.fixOutputSteering = max(self.fixOutputSteering-self.cfg.ROBOCARSHAT_OUTPUT_STEERING_TRIM_INC,1000)
+                    mylogger.info("CtrlIn Fixed output steering set to {}".format(self.fixOutputSteering))
+                self.sensor.setFixSteering (self.fixOutputSteering)            
+
+        if self.cfg.ROBOCARSHAT_STEERING_FIX != None:
+            user_steering = self.cfg.ROBOCARSHAT_STEERING_FIX
+
+        if (self.mode=='user' and self.cfg.ROBOCARSHAT_THROTTLE_FLANGER != None) :
+            user_throttle = self.dualMap(user_throttle,
+                -1, 0, 1,
+                self.cfg.ROBOCARSHAT_THROTTLE_FLANGER[0], 0, self.cfg.ROBOCARSHAT_THROTTLE_FLANGER[1])
+
+        if (self.mode=='user' and self.cfg.ROBOCARSHAT_THROTTLE_DISCRET != None) :
+            inds = np.digitize(user_throttle, self.discretesThrottle)
+            inds = max(inds,1)
+            user_throttle = self.cfg.ROBOCARSHAT_THROTTLE_DISCRET[inds-1]
+
+        #if switching back to user, then apply brake
+        if self.mode=='user' and self.lastMode != 'user' :
+            self.applyBrake=10 #brake duration
+
+        self.lastMode = self.mode
+        self.lastAux1 = self.inAux1
+        self.lastAux2 = self.inAux2
+        
+        if self.applyBrake>0:
+            user_throttle = self.cfg.ROBOCARSHAT_LOCAL_ANGLE_BRAKE_THROTTLE
+            self.applyBrake-=1
+
+        return user_throttle, user_steering
+
+    def update(self):
+
+        while self.on:
+            start = datetime.now()
+            self.getCommand()
+            stop = datetime.now()
+            s = 0.01 - (stop - start).total_seconds()
+            if s > 0:
```

Comment:
In general, you don't want to have any delay in your threaded loop, to give the part as little latency as possible. Is there any reason why you want to introduce a 10ms lag into the part?